File: backend/app/schemas/proposal.py
"""
Kaasb Platform - Proposal Schemas
Pydantic models for proposal validation and responses.
"""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

# === Nested Info (embedded in proposal responses) ===


class ProposalFreelancerInfo(BaseModel):
    """Minimal freelancer info shown on proposals."""

    id: uuid.UUID
    username: str
    first_name: str
    last_name: str
    display_name: str | None = None
    avatar_url: str | None = None
    title: str | None = None
    country: str | None = None
    experience_level: str | None = None
    skills: list[str] | None = None
    avg_rating: float = 0.0
    total_reviews: int = 0
    jobs_completed: int = 0

    model_config = {"from_attributes": True}


class ProposalJobInfo(BaseModel):
    """Minimal job info shown on freelancer's proposals list."""

    id: uuid.UUID
    title: str
    category: str
    job_type: str
    budget_min: float | None = None
    budget_max: float | None = None
    fixed_price: float | None = None
    status: str

    model_config = {"from_attributes": True}


# === Create / Update ===


class ProposalCreate(BaseModel):
    """Schema for submitting a proposal."""

    cover_letter: str = Field(min_length=50, max_length=5000)
    bid_amount: float = Field(ge=5, le=100000)
    estimated_duration: str | None = Field(None, max_length=50)


class ProposalUpdate(BaseModel):
    """Schema for freelancer updating their own proposal (while pending)."""

    cover_letter: str | None = Field(None, min_length=50, max_length=5000)
    bid_amount: float | None = Field(None, ge=5, le=100000)
    estimated_duration: str | None = Field(None, max_length=50)


class ProposalRespond(BaseModel):
    """Schema for client responding to a proposal (shortlist/accept/reject)."""

    status: str = Field(pattern=r"^(shortlisted|accepted|rejected)$")
    client_note: str | None = Field(None, max_length=2000)


# === Response Models ===


class ProposalDetail(BaseModel):
    """Full proposal detail — seen by the freelancer or the job's client."""

    id: uuid.UUID
    cover_letter: str
    bid_amount: float
    estimated_duration: str | None = None
    status: str
    client_note: str | None = None
    submitted_at: datetime
    responded_at: datetime | None = None
    created_at: datetime

    # Embedded relations
    freelancer: ProposalFreelancerInfo
    job: ProposalJobInfo

    model_config = {"from_attributes": True}


class ProposalSummary(BaseModel):
    """Proposal card — used in listings."""

    id: uuid.UUID
    bid_amount: float
    estimated_duration: str | None = None
    status: str
    submitted_at: datetime

    freelancer: ProposalFreelancerInfo
    job: ProposalJobInfo

    model_config = {"from_attributes": True}


class ProposalListResponse(BaseModel):
    """Paginated proposal list."""

    proposals: list[ProposalSummary]
    total: int
    page: int
    page_size: int
    total_pages: int

File: frontend/src/types/proposal.ts
export interface ProposalFreelancerInfo {
  id: string;
  username: string;
  first_name: string;
  last_name: string;
  display_name: string | null;
  avatar_url: string | null;
  title: string | null;
  country: string | null;
  experience_level: string | null;
  skills: string[] | null;
  avg_rating: number;
  total_reviews: number;
  jobs_completed: number;
}

export interface ProposalJobInfo {
  id: string;
  title: string;
  category: string;
  job_type: string;
  budget_min: number | null;
  budget_max: number | null;
  fixed_price: number | null;
  status: string;
}

export interface ProposalDetail {
  id: string;
  cover_letter: string;
  bid_amount: number;
  estimated_duration: string | null;
  status: string;
  client_note: string | null;
  submitted_at: string;
  responded_at: string | null;
  created_at: string;
  freelancer: ProposalFreelancerInfo;
  job: ProposalJobInfo;
}

export interface ProposalSummary {
  id: string;
  bid_amount: number;
  estimated_duration: string | null;
  status: string;
  submitted_at: string;
  freelancer: ProposalFreelancerInfo;
  job: ProposalJobInfo;
}

export interface ProposalListResponse {
  proposals: ProposalSummary[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export const PROPOSAL_STATUS_LABELS: Record<string, string> = {
  pending: "Pending",
  shortlisted: "Shortlisted",
  accepted: "Accepted",
  rejected: "Rejected",
  withdrawn: "Withdrawn",
};

export const PROPOSAL_STATUS_COLORS: Record<string, string> = {
  pending: "bg-yellow-50 text-yellow-700 border-yellow-200",
  shortlisted: "bg-blue-50 text-blue-700 border-blue-200",
  accepted: "bg-green-50 text-green-700 border-green-200",
  rejected: "bg-red-50 text-red-700 border-red-200",
  withdrawn: "bg-gray-100 text-gray-500 border-gray-200",
};
